### PR TITLE
Sagemaker inference agent

### PR DESCRIPTION
**Pull request by kumare3**
## Tracking issue
Fixes https://github.com/flyteorg/flyte/issues/4079

## Why are the changes needed?

To deploy models and generate predictions using Sagemaker on Flyte seamlessly.

## What changes were proposed in this pull request?

The Sagemaker agent lets users deploy models on Sagemaker, configure and invoke endpoints. 

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

**Comment by Future-Outlier** [comment]
If this PR needs my help, feel free to ping me, thanks a lot!

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker/workflow.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    wf.add_entity(
+        endpoint_config_task,
+        inputs=wf.inputs["endpoint_config_inputs"],
+        additional_args=endpoint_config_additional_args,
+    )
+
+    wf.add_entity(endpoint_task, inputs=wf.inputs["endpoint_inputs"])
+
+    lp = LaunchPlan.get_or_create(
```
Comment (kumare3) [inline plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker/workflow.py:67]: why return a launchplan? why not a workflow directly? we could register just once right and re-use!

File: plugins/flytekit-greatexpectations/setup.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 plugin_requires = [
     "flytekit>=1.5.0,<2.0.0",
-    "great-expectations>=0.13.30",
+    "great-expectations>=0.13.30,<=0.18.8",
```
Comment (eapolinario) [inline plugins/flytekit-greatexpectations/setup.py:9]: Can you separate this fix in its own PR?

File: flytekit/core/tracker.py [hunk 3]
```diff
@@ -318,7 +328,12 @@ def extract_task_module(f: Union[Callable, TrackedInstance]) -> Tuple[str, str,
         elif f.instantiated_in:
             mod = importlib.import_module(f.instantiated_in)
             mod_name = mod.__name__
-            name = f.lhs
+            try:
+                name = f.lhs
+            except Exception:
+                if not isPythonInstanceTask(f):
+                    raise AssertionError(f"Unable to determine module of {f}")
+                name = ""
```
Comment (eapolinario) [inline flytekit/core/tracker.py:336]: why do we need this?

File: flytekit/core/tracker.py [hunk 4]
```diff
@@ -305,6 +305,16 @@ def _task_module_from_callable(f: Callable):
     return mod, mod_name, name
 
 
+def isPythonInstanceTask(obj):
```
[the remark this answers is not in the document]
Comment (samhita-alla) [reply flytekit/core/tracker.py:308]: my bad. fixed it.

**Comment by eapolinario** [comment]
Just following up on the feedback we received last week, we should rename the new plugin to include the suffix `-serving` or something along those lines to indicate the capabilities of this new agent. Also, we should remove the current plugin as it's using a deprecated version of sagemaker (that depends on the sagemaker operator which hit EOL in [Feb-2023](https://github.com/aws/amazon-sagemaker-operator-for-k8s)).

**Comment by samhita-alla** [comment]
@eapolinario modified the plugin name to "flytekitplugins-awssagemaker-inference".

**Review by Future-Outlier — COMMENTED** [review]
Can we merge the latest branch?
It will solve some existed errors this PR.

File: plugins/flytekit-awssagemaker-inference/flytekitplugins/awssagemaker_inference/agent.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                    "result": TypeEngine.to_literal(
+                        ctx,
+                        json.dumps(endpoint_status, cls=DateTimeEncoder),
+                        str,
+                        TypeEngine.to_literal_type(str),
+                    )
+                }
+            )
+
+        return Resource(phase=flyte_state, outputs=res, message=message)
```
Comment (Future-Outlier) [inline plugins/flytekit-awssagemaker-inference/flytekitplugins/awssagemaker_inference/agent.py:101]: in flyte, `phase` is for flytepropeller and `state` is for flyteadmin.
so you have to use phase, cause flyteplugin is in flytepropeller.
```suggestion
        current_state = endpoint_status.get("EndpointStatus")
        flyte_phase = convert_to_flyte_phase(states[current_state])

        message = None
        if current_state == "Failed":
            message = endpoint_status.get("FailureReason")

        res = None
        if current_state == "InService":
            ctx = FlyteContextManager.current_context()
            res = LiteralMap(
                {
                    "result": TypeEngine.to_literal(
                        ctx,
                        json.dumps(endpoint_status, cls=DateTimeEncoder),
                        str,
                        TypeEngine.to_literal_type(str),
                    )
                }
            )

        return Resource(phase=flyte_phase, outputs=res, message=message)
```

File: flytekit/core/tracker.py [hunk 6]
```diff
@@ -89,6 +89,16 @@ def __call__(cls, *args, **kwargs):
         return o
 
 
+def is_python_instance_task(obj):
```
[the remark this answers is not in the document]
Comment (samhita-alla) [reply flytekit/core/tracker.py]: not necessary now. removed the modifications.

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/agent.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            res = LiteralMap(
+                {
+                    "result": TypeEngine.to_literal(
+                        ctx,
+                        json.dumps(endpoint_status, cls=DateTimeEncoder),
+                        str,
+                        TypeEngine.to_literal_type(str),
+                    )
+                }
+            )
```
Comment (pingsutw) [inline plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/agent.py:101]: nit: We can also return a dict; the agent service will convert the Python value to literal for you.
https://github.com/flyteorg/flytekit/blob/4a58a676717242fab2d3c9b94aabcc8a8116ffe7/flytekit/extend/backend/base_agent.py#L90
```suggestion
            res = {"result": json.dumps(endpoint_status, cls=DateTimeEncoder)}
```

File: plugins/flytekit-aws-sagemaker/setup.py [hunk 8]
```diff
@@ -27,11 +28,12 @@
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
         "Programming Language :: Python :: 3.10",
+        "Programming Language :: Python :: 3.11",
```
Comment (pingsutw) [inline plugins/flytekit-aws-sagemaker/setup.py:31]: Does this plugin work with Python 3.12?

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            outputs = LiteralMap(
+                {
+                    "result": TypeEngine.to_literal(
+                        ctx,
+                        result,
+                        dict,
+                        TypeEngine.to_literal_type(dict),
+                    )
+                }
+            )
```
Comment (pingsutw) [inline plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py:75]: ditto
```suggestion
            outputs = {"result": result}
```

**Review by pingsutw — COMMENTED** [review]
Overall LGTM, leave some minor comments

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/agent.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+        await self._call(
+            method="create_endpoint",
+            config=config,
+            inputs=inputs,
+            region=region,
+            aws_access_key_id=get_agent_secret(secret_key="aws-access-key"),
+            aws_secret_access_key=get_agent_secret(secret_key="aws-secret-access-key"),
+            aws_session_token=get_agent_secret(secret_key="aws-session-token"),
+        )
```
Comment (Future-Outlier) [inline plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/agent.py:68]: Should we change the secret to  `flyte_xxxx_xxxx_xxxx` ?
For example, `flyte_openai_access_token `?
cc @pingsutw

File: plugins/flytekit-aws-sagemaker/setup.py (hunk 8 above)
Comment (samhita-alla) [reply plugins/flytekit-aws-sagemaker/setup.py:31]: should work. updated the setup.py file. let me know if that works.

**Review by pingsutw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
